# Temaøvelse 2

**Algebraens fundamentalsætning** siger, at ethvert polynomium $ p(Z) \in \mathbb{C}[Z] $ af grad mindst én har en rod $ \lambda \in \mathbb{C} $ (se Sætning 4.6.1 i lærebogen). For praktiske anvendelser er det vigtigt at kunne finde en sådan rod eller i det mindste at finde en god numerisk tilnærmelse, også kaldet en approksimation.

Målet med denne temaøvelse er at opnå algoritmisk indsigt i, hvordan numeriske approksimationer af rødderne i polynomier med reelle koefficienter kan findes. Vi vil gennemgå to metoder: **bisektionsmetoden** og **Newton-Raphson-metoden**.

## Del I: Bisektionsmetoden

Begynd med at åbne en kommandolinjeversion af Python. (Vi bruger ipynb)

## 1. 
For at kunne arbejde med rødderne i polynomier $p(X) \in \mathbb{R}[X]$, definerer vi funktionen, som et sådant polynomium giver anledning til: $p : \mathbb{R} \to \mathbb{R}$, hvor $x \mapsto p(x)$. Det første skridt er at finde ud af, hvordan man definerer en sådan funktion i Python.

### (a)

Overvej som et eksempel polynomiet $p(X) = X^3 - X - 6$. Den tilsvarende funktion $p : \mathbb{R} \to \mathbb{R}$ kan defineres i Python som følger:

```python
def p(x):
    return x**3 - x - 6


In [2]:
def p(x):
    return x**3 - x - 6

### (b)

Nu kan du beregne værdierne af funktionen $p$ ved hjælp af Python. Skriv for eksempel $p(10)$ og kontroller i hånden, at Python returnerer den korrekte værdi.


In [3]:
p(10) # 10^3 - 10 - 6 = 1000 - 16 = 984

984

### (c)

Brug Python til at beregne $p(-1)$, $p(0)$, $p(1)$, $p(2)$ og $p(3)$. Kan du nu angive en rod i polynomiet $p(X)$?


In [4]:
print(f"p(-1) = {p(-1)}")
print(f"p(0) = {p(0)}")
print(f"p(1) = {p(1)}")
print(f"p(2) = {p(2)}")
print(f"p(3) = {p(3)}")

p(-1) = -6
p(0) = -6
p(1) = -6
p(2) = 0
p(3) = 18


Generelt vil man sjældent være heldig at finde en rod i et polynomium $f(X)$ blot ved at beregne nogle enkelte af dens værdier. Derfor vil vi nu studere et kriterium, der af og til kan anvendes til i det mindste at bestemme, hvor en rod cirka er placeret. Vi vil starte med at formulere en sætning for kontinuerte funktioner. Vi vil ikke definere, hvad en kontinueret funktion præcist er, men meget løst sagt er en funktion kontinueret, hvis dens graf ikke har nogen “huller”. 

Du må frit benytte i denne temaøvelse, at enhver funktion $f : \mathbb{R} \to \mathbb{R}$, hvor $x \mapsto f(x)$, og hvor $f(X) \in \mathbb{R}[X]$ er et polynomium, er kontinueret. Kontinuerte funktioner har nogle rare egenskaber, blandt andet den følgende sætning, kendt som mellemværdisætningen.

### Sætning 1 
Lad $a, b$ være reelle tal, der opfylder $a < b$. Lad yderligere $f : [a,b] \to \mathbb{R}$ være en kontinueret funktion.

- Hvis $f(a) < f(b)$, og $y$ opfylder $f(a) < y < f(b)$, så er $y = f(x)$ for nogle $x$ i intervallet $[a, b]$.

- Hvis $f(a) > f(b)$, og $y$ opfylder $f(b) < y < f(a)$, så er $y = f(x)$ for nogle $x$ i intervallet $[a, b]$.

Vi vil ikke bevise denne sætning, men du må bruge den frit. Hvis $f : \mathbb{R} \to \mathbb{R}$ er en funktion, og $\lambda \in \mathbb{R}$ opfylder $f(\lambda) = 0$, kalder vi $\lambda$ for et nulpunkt i $f$. Hvis $f$ er defineret ud fra et polynomium $f(X)$, er et nulpunkt i $f$ simpelthen en rod i polynomiet $f(X)$. Mellemværdisætningen leder til en nyttig konsekvens vedrørende nulpunkter i kontinuerte funktioner:

### Korollar 2 
Lad $a, b$ være reelle tal, der opfylder $a < b$. Lad yderligere $f : [a,b] \to \mathbb{R}$ være en kontinueret funktion, der opfylder $f(a) \cdot f(b) < 0$. Da eksisterer der $x \in [a,b]$, således at $f(x) = 0$. Med andre ord: funktionen $f$ har et nulpunkt i intervallet $[a, b]$.

## 2. 
Målet med dette spørgsmål er at forstå, hvorfor Korollar 2 er en logisk konsekvens af Sætning 1.


### (a), (b) og (c)
Vis først og fremmest, at $f(a) \cdot f(b) < 0$ medfører, at 
$f(a) < 0 \land f(b) > 0$ eller $f(a) > 0 \land f(b) < 0$.

### Svar
Korollar 2 følger logisk af Sætning 1, fordi hvis $f(a) \cdot f(b) < 0$, så betyder det, at $f(a)$ og $f(b)$ har forskellige fortegn. Det vil sige, at enten $f(a) < 0$ og $f(b) > 0$, eller $f(a) > 0$ og $f(b) < 0$. Ifølge Sætning 1 skal der derfor eksistere en værdi $x$ mellem $a$ og $b$, hvor $f(x) = 0$, hvilket er præcis, hvad Korollar 2 siger.

Hvis vi antager, at $f(a) \cdot f(b) < 0$, så må enten $f(a) < 0$ og $f(b) > 0$, eller $f(a) > 0$ og $f(b) < 0$, da produktet af to tal med forskellige fortegn altid vil være negativt. Dette er også i overensstemmelse med Sætning 1, som bekræfter, at der må være et punkt $x$ i intervallet $[a, b]$, hvor $f(x) = 0$, når der er et skift i fortegn.


## 3.
Polynomiet $p(X) = X^3 - X - 6$ blev med vilje valgt på en sådan måde, at det havde den “pæne” rod 2. Vi vælger nu et andet polynomium uden sådanne pæne rødder.
### (a)
Betragt polynomiet $f(X) = X^3 - X - 5$, og definér i Python den tilsvarende funktion $f : \mathbb{R} \to \mathbb{R}$, det vil sige funktionen, der opfylder $x \mapsto x^3 - x - 5$.

In [5]:
def f(x):
    return x**3 - x - 5

### (b)
Benyt Python til at udregne $f(x)$ for $x \in \{0, 1, 2, 3\}$. Konkludér ved hjælp af Korollar 2 og det faktum, at $f(1) < 0$ og $f(2) > 0$, at polynomiet $f(X) = X^3 - X - 5$ har en rod i intervallet $[1, 2]$. Som nævnt tidligere må du frit benytte, at en polynomiefunktion er kontinueret.


In [6]:
for i in range(4):
    print(f"f({i}) = {f(i)}")

f(0) = -5
f(1) = -5
f(2) = 1
f(3) = 19


#### Konklusion
Ved hjælp af Python har vi beregnet værdierne af $f(x)$ for $x \in \{0, 1, 2, 3\}$ for polynomiet $f(X) = X^3 - X - 5$. Resultaterne viser følgende:

- $f(0) = -5$
- $f(1) = -5$
- $f(2) = 1$
- $f(3) = 19$

Vi observerer, at $f(1) < 0$ og $f(2) > 0$. Ifølge Korollar 2, og fordi polynomier er kontinuerte, må der eksistere en rod i intervallet $[1, 2]$, hvor $f(x) = 0$. Dette betyder, at polynomiet $f(X) = X^3 - X - 5$ har en rod mellem $x = 1$ og $x = 2$.

### (c)
For et givet interval $[a,b]$ kalder man værdien $\frac{(a+b)}{2}$ for midtpunktet af dette interval og værdien $b - a$ for bredden af intervallet. For eksempel har intervallet $[1, 2]$ midtpunktet $\frac{(1 + 2)}{2} = 1.5$ og bredden $2 - 1 = 1$. Udregn nu $f(x)$ i midtpunktet af intervallet $[1, 2]$. Har polynomiet $X^3 - X - 5$ en rod i intervallet $[1, 1.5]$ eller i intervallet $[1.5, 2]$? Bemærk, at disse intervaller har bredden $1/2$, det vil sige halvdelen af bredden af intervallet $[1, 2]$.


In [7]:
print(f"f(1.5) = {f(1.5)}")

f(1.5) = -3.125


#### Svar
Ved at udregne $f(1.5)$ får vi værdien $f(1.5) = -3.125$. Da $f(1.5) < 0$ og $f(2) > 0$ fra den tidligere udregning, kan vi ifølge Korollar 2 konkludere, at der eksisterer en rod i intervallet $[1.5, 2]$, fordi funktionen skifter fortegn inden for dette interval. Dette betyder, at polynomiet $X^3 - X - 5$ har en rod mellem $x = 1.5$ og $x = 2$.

### (d)
Bestem nu, baseret på værdien af $f(x)$ i midtpunktet af det interval, du lige har
fundet, et interval med bredden $\frac{1}{4}$ indeholdende en rod i polynomiet $X^3 − X − 5$.

#### Svar
Vi har tidligere fundet, at roden af polynomiet $X^3 - X - 5$ ligger i intervallet $[1.5, 2]$. Nu skal vi finde midtpunktet af dette interval for at kunne indsnævre roden yderligere.

Midtpunktet af intervallet $[1.5, 2]$ er:

In [8]:
(1.5 + 2) / 2

1.75

Vi udregner nu $f(1.75)$:

In [9]:
print(f"f(1.75) = {f(1.75)}")

f(1.75) = -1.390625


Vi ser at $f(1.75)$ er en værdi, der er mindre end 0, og vi ved, at $f(2) > 0$. Det betyder, at roden ligger i intervallet $[1.75, 2]$. Vi har nu at bredden er lig med:

In [10]:
2 - 1.75

0.25

Vi har fundet at intervallet $[1.75, 2]$ har bredden $\frac{1}{4}$.

### (e)
Man kan i princippet fortsætte proceduren fra 3 (d) adskillige gange, hvor man hver gang finder et interval, der indeholder en rod i $X^3 - X - 5$, med en bredde, der er halvt så stor som det forrige intervals. Udfør yderligere to trin, og bestem et interval med bredden $1/16$, der indeholder en rod i polynomiet $X^3 - X - 5$.


#### Svar
Vi udregner de næste to midtpunkter.

In [11]:
(1.75 + 2) / 2

1.875

In [12]:
(1.875 + 2) / 2

1.9375

Vi tjekker nu om vi holder reglen $f(x) < 0$ og $f(x_1) > 0$

In [13]:
print(f"f(1.875) = {f(1.875)}")
print(f"f(1.9375) = {f(1.9375)}")

f(1.875) = -0.283203125
f(1.9375) = 0.335693359375


Da $f(1.875) < 0$ og $f(1.9375) > 0$, er roden i intervallet $[1.875, 1.9375]$. Vi udregner bredden til at være:

In [14]:
1.9375 - 1.875

0.0625

##### Konklusion:
Efter yderligere to trin har vi indsnævret roden til intervallet $[1.875, 1.9375]$ med en bredde på $1/16$, som indeholder en rod i polynomiet $X^3 - X - 5$.

Idéerne, der blev præsenteret herover, giver anledning til en algoritme, der kan finde approksimationer af rødderne i et polynomium med reelle koefficienter. Mere præcist: givet $p(X) \in \mathbb{R}[x]$ og $a, b \in \mathbb{R}$ sådan at $a < b$, og $p(a) \cdot p(b) < 0$, så kan vi rekursivt definere en sekvens af intervaller $[a_0, b_0], [a_1, b_1], [a_2, b_2], \dots$ som følger:

$
[a_n, b_n] = 
\begin{cases} 
    [a, b] & \text{hvis } n = 0 \\
    \left[ a_{n-1}, \frac{a_{n-1} + b_{n-1}}{2} \right] & \text{hvis } n \geq 1 \text{ og } p(a_{n-1}) \cdot p\left(\frac{a_{n-1} + b_{n-1}}{2}\right) \leq 0 \\
    \left[ \frac{a_{n-1} + b_{n-1}}{2}, b_{n-1} \right] & \text{hvis } n \geq 1 \text{ og } p(a_{n-1}) \cdot p\left(\frac{a_{n-1} + b_{n-1}}{2}\right) > 0.
\end{cases}
$

At benytte midtpunkterne i alle de fundne intervaller giver derefter bedre og bedre approksimationer af en rod i polynomiet $p(X)$. Med andre ord: sekvensen af reelle tal $r_0, r_1, r_2, \dots$ defineret som

$
r_n = \frac{a_n + b_n}{2}
$

er approksimationer af en rod i $p(X)$. Jo større værdien af $n$ er, desto bedre bliver approksimationen. Denne metode til tilnærmelse af rødder kaldes bisektionsmetoden.


## 4.
### (a)
Benyt Python på polynomiet $f(X) = X^3 - X - 5$ med $[a, b] = [1, 2]$ til at bestemme $r_4$. Du kan genbruge de intervaller, du allerede har bestemt i den tidligere opgave. Benyt også Python til at udregne $f(r_4)$. Svar: $r_4 = 1.90625$, og $f(r_4) = 0.020660400390625$.

### Svar
For at løse denne opgave opstiller vi funktionen ovenfor.

In [15]:
from typing import Callable

def approxRoot(function: Callable[[float], float], a: float, b: float, n: int = 0):
    if n == 0:
        a,b = a, b
        r = (a + b)/2
        x = function(r)

        out = f'''
        a = {a}
        b = {b}
        r = {r}
        function(r) = {x}
        '''
        print(out)
        return
    elif n >= 1 and function(a) * function((a + b)/2) <= 0:
        b = (a + b)/2
    elif n >= 1 and function(a) * function((a + b)/2) > 0:
        a = (a + b)/2
    approxRoot(function, a, b, n - 1)

In [16]:
approxRoot(f,a=1,b=2,n=4)


        a = 1.875
        b = 1.9375
        r = 1.90625
        function(r) = 0.020660400390625
        


### (b)
Antag fortsat, at $f(X) = X^3 - X - 5$, og $[a, b] = [1, 2]$, og vis ved induktion på $n$, at bredden af intervallet $[a_n, b_n]$ er lig med $1 / 2^n$ for alle $n \in \mathbb{Z}_{>0}$. En konsekvens af dette er, at afstanden fra $r_n$ til en rod i $f(X)$ højst er $1 / 2^{n+1}$.

#### Induktionsstart
For n = 0 er bredden af det første interval $[a_0, b_0] = [1,2]$, så $\newline$
$b_0 - a_0 = 2 - 1 = 1 \newline$ 

Vi bemærker nu at dette er lig med $\frac{1}{20}$, som stemmer overens med påstanden. Derfor gælder påstanden for $n = 0$.

#### Induktionsantagelse
Antag, at påstanden gælder for $n = k$, altså at bredden af intervallet $[a_k, b_k]$ er: $\newline$
$b_k - a_k = \frac{1}{2^k}$

#### Induktionsskridt
Vi skal nu vise, at påstanden også gælder for $ n = k + 1 $. Det vil sige, vi skal vise, at bredden af intervallet $ [a_{k+1}, b_{k+1}] $ er $ \frac{1}{2^{k+1}} $.

Ved bisektionsmetoden bliver det nye interval $ [a_{k+1}, b_{k+1}] $ fundet ved at halvere bredden af intervallet $ [a_k, b_k] $. Det betyder:

$
b_{k+1} - a_{k+1} = \frac{b_k - a_k}{2}
$

Indsætter vi induktionsantagelsen, får vi:

$
b_{k+1} - a_{k+1} = \frac{1}{2} \cdot \frac{1}{2^k} = \frac{1}{2^{k+1}}
$

Dermed er påstanden også sand for $ n = k + 1 $.

#### Afstand fra $r_n$ til roden
Midtpunktet $r_n$ af intervallet $[a_n, b_n]$ er givet ved:

$$
r_n = \frac{a_n + b_n}{2}
$$

Da vi har vist, at bredden af intervallet $[a_n, b_n]$ er $\frac{1}{2^n}$, og midtpunktet er midt i dette interval, er afstanden fra $r_n$ til en rod højst halvdelen af intervalbredden:

$$
\text{Afstand} = \frac{1}{2} \cdot \frac{1}{2^n} = \frac{1}{2^{n+1}}
$$


### Bemærkning 3
Bemærk som en lille advarsel, at man i en realistisk anvendelse måske kun kender koefficienterne af $p(X)$ til en vis numerisk præcision. Hvis $(a_{n-1} + b_{n-1})/2$ allerede er tæt på en rod i $p(X)$, kan det derfor være umuligt pålideligt at bestemme, om $p(a_{n-1}) \cdot p\left(\frac{a_{n-1} + b_{n-1}}{2}\right) > 0$ eller $p(a_{n-1}) \cdot p\left(\frac{a_{n-1} + b_{n-1}}{2}\right) \leq 0$, og man kan ikke være sikker på andet end, at $p(a_{n-1}) \cdot p\left(\frac{a_{n-1} + b_{n-1}}{2}\right)$ er meget tæt på nul. Sådanne overvejelser er emner for numerisk analyse og er vigtige, når man udvikler såkaldt numeriske algoritmer.


## Del II: Newton-Raphsons metode til bestemmelse af rødder i polynomier med reelle koefficienter

Hvis alt gik godt, fandt du i det forrige, at polynomiet $X^3 - X - 5$ har en rod, der er indeholdt i intervallet $[1.875, 1.9375]$. Midtpunktet i dette interval er derfor en rimelig approksimation for en rod i $X^3 - X - 5$. Som allerede udregnet i den forrige opgave er dette midtpunkt lig med $\frac{(1.875 + 1.9375)}{2} = 1.90625$. Vi vil nu behandle en anden måde, hvorpå man kan finde numeriske tilnærmelser af rødderne i et polynomium $p(X) \in \mathbb{R}[X]$.

## 5. 
Lad $p(X) \in \mathbb{R}[X]$ være et polynomium af grad mindst én. Desuden antages det, at $\lambda \in \mathbb{R}$ er en rod i $p(X)$. Som før definerer vi den funktion, som dette polynomium giver anledning til, ved $p : \mathbb{R} \to \mathbb{R}$, hvor $x \mapsto p(x)$.

### (a)
Antag, at $\lambda$ er en rod i $p(X)$ med multiplicitet $m$, hvor $m$ er mindst to. Benyt Definition 4.6.1 i lærebogen til at konkludere, at der i dette tilfælde eksisterer et polynomium $g(X)$, således at $f(X) = (X - \lambda)^m \cdot g(X)$. Benyt nu antagelsen om, at $m$ er mindst to, til at udlede, at der findes et polynomium $h(X)$, således at $f(X) = (X - \lambda)^2 \cdot h(X)$.


![Definition 4.6.1](../../../misc/images/definitions/4.6.1.png)


#### Svar:
Vi er blevet givet $\lambda$ som en rod i polynomiet $p(X)$ og det er angivet at multiplicitetten $m \geq 2$. Da vi ved at polynomiet kan skrives som: $\newline f(X) = (X - \lambda)^m \cdot h(X)$

Siden $m \geq 2$, ved vi, at $ (X - \lambda) $ mindst optræder to gange som en faktor i udtrykket. Derfor kan vi faktorisere $ (X - \lambda)^2 $ ud fra udtrykket for $ f(X) $. Dette giver:
$
\newline f(X) = (X - \lambda)^2 \cdot \underbrace{(X - \lambda)^{m-2} \cdot g(X)}_{h(X)}
$

Den resterende faktor efter at have faktoriseret $ (X - \lambda)^2 $ er $ (X - \lambda)^{m-2} \cdot g(X) $, som vi kalder $ h(X) $. Således definerer vi:
$
\newline h(X) = (X - \lambda)^{m-2} \cdot g(X)
$
Dette er også et polynomium, fordi både $ (X - \lambda)^{m-2} $ og $ g(X) $ er polynomier.

Vi har nu vist, at der eksisterer et polynomium $ h(X) $, således at:
$
\newline f(X) = (X - \lambda)^2 \cdot h(X)
$
hvor $ h(X) = (X - \lambda)^{m-2} \cdot g(X) $.


### (b)
For et givet polynomium

$
\newline f(X) = a_0 + a_1 X + a_2 X^2 + \cdots + a_n X^n
$

defineres den afledte af $ f(X) $, betegnet ved $ f'(X) $ eller sommetider også $ f(X)' $, på sædvanlig måde som

$
\newline f'(X) = a_1 + 2a_2 X + \cdots + n a_n X^{n-1}.
$

Vis, at hvis $ \lambda $ er en rod i $ f(X) $ med multiplicitet mindst to, så er $ \lambda $ en rod i $ f'(X) $. Hint: udnyt, at

$
\newline f(X) = (X - \lambda)^2 \cdot h(X)
$

for et polynomium $ h(X) $, og benyt produktreglen for differentiation.


### Svar:
Vi er givet, at $ f(X) $ kan skrives som:

$
\newline f(X) = (X - \lambda)^2 \cdot h(X)
$

hvor $ h(X) $ er et polynomium. Dette betyder, at $ \lambda $ er en rod af $ f(X) $ med multiplicitet mindst to, og vi skal nu vise, at $ \lambda $ også er en rod af $ f'(X) $.

#### Trin 1: Differentiering af $ f(X) $ ved brug af produktreglen

For at differentiere $ f(X) = (X - \lambda)^2 \cdot h(X) $, benytter vi produktreglen:

$
\newline (f \cdot g)' = f' \cdot g + f \cdot g'
$

Anvend produktreglen på $ f(X) = (X - \lambda)^2 \cdot h(X) $:

$
\newline f'(X) = \frac{d}{dX} \left( (X - \lambda)^2 \right) \cdot h(X) + (X - \lambda)^2 \cdot h'(X)
$

Den afledte af $ (X - \lambda)^2 $ er:

$
\newline \frac{d}{dX} \left( (X - \lambda)^2 \right) = 2(X - \lambda)
$

Så vi får:

$
\newline f'(X) = 2(X - \lambda) \cdot h(X) + (X - \lambda)^2 \cdot h'(X)
$

#### Trin 2: Sæt $ X = \lambda $

Vi skal nu undersøge, om $ \lambda $ er en rod af $ f'(X) $. For at gøre det, indsætter vi $ X = \lambda $ i udtrykket for $ f'(X) $:

$
\newline f'(\lambda) = 2(\lambda - \lambda) \cdot h(\lambda) + (\lambda - \lambda)^2 \cdot h'(\lambda)
$

Bemærk, at begge led indeholder faktorer af $ (\lambda - \lambda) = 0 $, så begge led bliver nul:

$
\newline f'(\lambda) = 2(0) \cdot h(\lambda) + 0^2 \cdot h'(\lambda) = 0
$

Derfor er $ f'(\lambda) = 0 $, hvilket betyder, at $ \lambda $ er en rod i $ f'(X) $.

### (c)
Antag, at $ \lambda $ er en rod i $ f(X) $ med multiplicitet én. Vis, at der i dette tilfælde eksisterer et polynomium $ g(X) $, således at

$
\newline f(X) = (X - \lambda) \cdot g(X)
$

og

$
\newline g(\lambda) \neq 0.
$

### (d)
Vis, at hvis $ \lambda $ er en rod i $ f(X) $ med multiplicitet én, så er $ f'(\lambda) \neq 0 $. Hint: benyt igen produktreglen til at finde et udtryk for $ f'(X) $, denne gang på produktet 

$
\newline f(X) = (X - \lambda) \cdot g(X).
$

Udnyt derefter det faktum, at $ g(\lambda) \neq 0 $.

#### Svar til (c) og (d):
Ligesom i forrige opgave ved vi, at hvis $ \lambda $ er en rod i $ f(X) $ med multiplicitet én, betyder det, at $ f(X) $ kan faktoriseres som $ (X - \lambda) $. Vi har nu, at:

$
\newline f(X) = (X - \lambda) \cdot g(X)
$

hvor $ g(X) $ ikke har $ \lambda $ som en rod. Med andre ord, $ g(\lambda) \neq 0 $.

For at verificere dette, kan vi anvende produktreglen til at differentiere $ f(X) $:

$
\newline f'(X) = \frac{d}{dX} \left( (X - \lambda) \cdot g(X) \right)
$

Produktreglen siger, at:

$
\newline (f \cdot g)' = f' \cdot g + f \cdot g'
$

Anvend produktreglen på udtrykket for $ f(X) $:

$
\newline f'(X) = \frac{d}{dX} \left( X - \lambda \right) \cdot g(X) + (X - \lambda) \cdot g'(X)
$

Da den afledte af $ X - \lambda $ er 1, får vi:

$
\newline f'(X) = g(X) + (X - \lambda) \cdot g'(X)
$

Når vi sætter $ X = \lambda $, får vi:

$
\newline f'(\lambda) = g(\lambda) + (\lambda - \lambda) \cdot g'(\lambda)
$

Da $ (\lambda - \lambda) = 0 $, forsvinder det andet led, og vi står tilbage med:

$
\newline f'(\lambda) = g(\lambda)
$

Da vi ved, at $ g(\lambda) \neq 0 $, følger det, at $ f'(\lambda) \neq 0 $.

### Konklusion:
Vi har nu vist, at hvis $ \lambda $ er en rod i $ f(X) $ med multiplicitet én, så er $ f'(\lambda) \neq 0 $, som vi ønskede at bevise.


# 7

In [26]:
def newtonRaphsons(function: Callable[[float], float], prime_function: Callable[[float], float], n: int = 0, x0: float = 0) -> float:
    if n==0:
        return x0
    else:
        return (newtonRaphsons(function,prime_function, n - 1, x0)-function(newtonRaphsons(function,prime_function, n - 1, x0))/prime_function(newtonRaphsons(function,prime_function, n - 1, x0)))

In [27]:
def f(x):
    return x**3 - x - 5

def fprime(x):
    return 3*x**2 - 1

In [34]:
newtonRaphsons(function=f, prime_function=fprime, n=2, x0=2)

1.9041748600816821

: 

1.9041748600816821